sentry-types/src/protocol/envelope.rs

```diff
@@ -1,16 +1,82 @@
-use std::io::Write;
+use std::{io::Write, path::Path};
 
+use serde::Deserialize;
+use thiserror::Error;
 use uuid::Uuid;
 
 use super::v7::{Attachment, Event, SessionAggregates, SessionUpdate, Transaction};
 
+/// Raised if a envelope cannot be parsed from a given input.
+#[derive(Debug, Error)]
+pub enum EnvelopeError {
+    /// Unexpected end of file
+    #[error("unexpected end of file")]
+    UnexpectedEof,
+    /// Missing envelope header
+    #[error("missing envelope header")]
+    MissingHeader,
+    /// Missing item header
+    #[error("missing item header")]
+    MissingItemHeader,
+    /// Missing newline after header or payload
+    #[error("missing newline after header or payload")]
+    MissingNewline,
+    /// Invalid envelope header
+    #[error("invalid envelope header")]
+    InvalidHeader(#[source] serde_json::Error),
+    /// Invalid item header
+    #[error("invalid item header")]
+    InvalidItemHeader(#[source] serde_json::Error),
+    /// Invalid item payload
+    #[error("invalid item payload")]
+    InvalidItemPayload(#[source] serde_json::Error),
+}
+
+#[derive(Deserialize)]
+struct EnvelopeHeader {
+    event_id: Option<Uuid>,
+}
+
+/// An Envelope Item Type.
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize)]
+enum EnvelopeItemType {
+    /// An Event Item type.
+    #[serde(rename = "event")]
+    Event,
+    /// A Session Item type.
+    #[serde(rename = "session")]
+    SessionUpdate,
+    /// A Session Aggregates Item type.
+    #[serde(rename = "sessions")]
+    SessionAggregates,
+    /// A Transaction Item type.
+    #[serde(rename = "transaction")]
+    Transaction,
+    /// An Attachment Item type.
+    #[serde(rename = "attachment")]
+    Attachment,
+}
+
+/// An Envelope Item Header.
+#[allow(dead_code)]
+#[derive(Clone, Debug, Deserialize)]
+struct EnvelopeItemHeader {
+    r#type: EnvelopeItemType,
+    length: Option<usize>,
+    // Fields below apply only to Attachment Item type
+    filename: Option<String>,
+    attachment_type: Option<String>,
+    content_type: Option<String>,
+}
+
 /// An Envelope Item.
 ///
 /// See the [documentation on Items](https://develop.sentry.dev/sdk/envelopes/#items)
 /// for more details.
-#[derive(Clone, Debug, PartialEq)]
+#[derive(Clone, Debug, Deserialize, PartialEq)]
 #[non_exhaustive]
 #[allow(clippy::large_enum_variant)]
+#[serde(untagged)]
```

Comment:
We don’t `Deserialize` this enum on itself, but rather the inner types directly, so we don’t need the annotations here.